extensions/amp-next-page/0.2/page.js

```diff
@@ -0,0 +1,195 @@
+    this.visibilityState_ = VisibilityState.PRERENDER;
+
+    // Public props
+    /** @type {string} */
+    this.title = title;
+    /** @type {string} */
+    this.url = url;
+    /** @type {string} */
+    this.image = image;
+    /** @type {!PageRelativePos} */
+    this.relativePos = PageRelativePos.OUTSIDE_VIEWPORT;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  isVisible() {
+    return this.visibilityState_ == VisibilityState.VISIBLE;
+  }
+
+  /**
+   * @return {!VisibilityState}
+   * @visibleForTesting
+   */
+  getVisibilityState() {
+    return this.visibilityState_;
+  }
+
+  /**
+   * @param {boolean} visible
+   */
+  setVisible(visible) {
+    // TODO(wassgha): Handle history manipulation
+    // TODO(wassgha): Handle manual visibility management
+    const visibilityState = visible
+      ? VisibilityState.VISIBLE
+      : VisibilityState.HIDDEN;
+
+    // Update visibility internally and at the shadow doc level
+    if (this.shadowDoc_ && visibilityState != this.visibilityState_) {
+      this.shadowDoc_.setVisibilityState(visibilityState);
+      this.visibilityState_ = visibilityState;
+    }
+  }
+
+  /**
+   * @return {boolean}
+   */
+  isFetching() {
+    return this.state_ == PageState.FETCHING;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  isLoaded() {
+    return this.state_ == PageState.LOADED || this.state_ == PageState.INSERTED;
+  }
+
+  /**
+   * Asks the next-page manager to fetch the document's HTML
+   * and inserts it
+   * @return {!Promise}
+   */
+  fetch() {
+    if (this.state_ == PageState.INSERTED) {
+      return Promise.resolve();
+    }
+
+    this.state_ = PageState.FETCHING;
+
+    return this.manager_
+      .fetchPageDocument(this)
+      .then(content => {
+        this.state_ = PageState.LOADED;
+
+        const shadowDoc = this.manager_.appendAndObservePage(this, content);
+        if (shadowDoc) {
+          this.shadowDoc_ = shadowDoc;
+          this.manager_.setLastFetchedPage(this);
+          this.state_ = PageState.INSERTED;
+        } else {
+          this.state_ = PageState.FAILED;
+        }
+      })
+      .catch(() => {
+        this.state_ = PageState.FAILED;
+      });
+  }
+
+  /**
+   * Called when a position change is detected on the injected
+   * header and footer elements
+   * @param {!PageBound} bound
+   * @param {?PositionInViewportEntryDef} position
+   */
+  boundPositionChanged(bound, position) {
+    const prevBoundPosition = this.boundPosition_[bound];
+    if (position.relativePos === prevBoundPosition) {
+      return;
+    }
+    this.boundPosition_[bound] = position.relativePos;
+    this.updateRelativePos_();
+  }
+
+  /**
+   * Calculates the position of the document relative to the viewport
+   * based on the positions of the injected footer and header elements
+   * @private
+   */
+  updateRelativePos_() {
+    const header = this.boundPosition_[PageBound.HEADER];
+    const footer = this.boundPosition_[PageBound.FOOTER];
+
```

Comment:
Maybe we should double check what this function is doing and clarify a bit if possible (e.g. what happens if `header` is `null`?).